### PR TITLE
std.Build: add --seed argument to randomize step dependencies spawning

### LINKED ISSUE #14940: accept a seed option to the build runner and shuffle the dependency order using it
Extracted from #14647.

Add a --seed option to the build runner and make the `zig build` subcommand always supply a random seed argument.

Use this seed to shuffle the dependency order of steps when -j1 is selected, and ideally also shuffle order of steps for dependency peers when concurrency is being used. Be sure not to disrupt the heuristics for starting leaf dependencies first when the number of threads in the thread pool is greater than one.

This will help protect build scripts from accumulating concurrency bugs.

### PR DESCRIPTION
**Pull request by Sahnvour**
Closes #14940

### CODE REVIEW

File: lib/build_runner.zig [hunk 1]
```diff
@@ -731,10 +745,22 @@ fn printTreeStep(
...
+/// Along the way, check that there is no dependency loop, and record the steps
+/// in traversal order in `step_stack`.
+/// Each step has its dependencies traversed in random order, this accomplishes
+/// two things:
+/// - `step_stack` will be in randomized-depth-first order, so the build runner
+///   spawns steps in a random (but optimized) order
+/// - each step's `dependants` list is also filled in a random order, so that
+///   when it finishes executing in `workerMakeOneStep`, it spawns next steps
+///   to run in random order
+fn buildAndCheckGraphForDependencyLoop(
```
Comment (andrewrk) [inline lib/build_runner.zig:759]: why was this renamed to "buildAnd..." ? It's not building yet.

File: lib/build_runner.zig [hunk 2]
```diff
@@ -369,8 +375,14 @@ fn runStepNames(
     }
 
     const starting_steps = try arena.dupe(*Step, step_stack.keys());
+
+    const useed = std.hash.XxHash64.hash(0x1312, b.seed);
+    var rng = std.rand.DefaultPrng.init(useed);
```
Comment (andrewrk) [inline lib/build_runner.zig:380]: how about instead accepting a 32-bit integer, possibly hex encoded? This is fundamentally simpler, and does not leave the user wondering what string they should use for the seed.

File: lib/build_runner.zig [hunk 3]
```diff
@@ -1014,6 +1046,7 @@ fn usage(builder: *std.Build, already_ran_build: bool, out_stream: anytype) !voi
         \\  --global-cache-dir [path]    Override path to global Zig cache directory
         \\  --zig-lib-dir [arg]          Override path to Zig lib directory
         \\  --build-runner [file]        Override path to build runner
+        \\  --seed [string]              Seed the prng used in the build runner
```
Comment (andrewrk) [inline lib/build_runner.zig:1049]:
```suggestion
        \\  --seed [string]              For shuffling dependency traversal order (default: 0)
```

File: lib/build_runner.zig [hunk 4]
```diff
@@ -192,6 +192,12 @@ pub fn main() !void {
                     std.debug.print("Expected argument after {s}\n\n", .{arg});
                     usageAndErr(builder, false, stderr_stream);
                 } };
+            } else if (mem.eql(u8, arg, "--seed")) {
+                const next_arg = nextArg(args, &arg_idx) orelse {
+                    std.debug.print("Expected a string after {s}\n\n", .{arg});
+                    usageAndErr(builder, false, stderr_stream);
+                };
+                builder.seed = next_arg;
```
Comment (andrewrk) [inline lib/build_runner.zig:200]: should use a local variable instead of storing extra state into Build

File: lib/std/Build.zig [hunk 5]
```diff
@@ -96,6 +96,10 @@ cache_root: Cache.Directory,
 global_cache_root: Cache.Directory,
 cache: *Cache,
 zig_lib_dir: ?LazyPath,
+// The hash of these bytes will be used as an init value for the build prng.
+// It allows receiving any input as seed, and printing it to output for
+// reproducibility.
+seed: []const u8,
```
Comment (andrewrk) [inline lib/std/Build.zig:102]: if you want to expose this to build scripts, let's make that a separate proposal.

**Review by andrewrk — CHANGES_REQUESTED** [review]
Thanks for working on this! 

Don't forget that part of the acceptance criteria is

> make the `zig build` subcommand always supply a random seed argument.

File: lib/build_runner.zig (hunk 1 above)
Comment (Sahnvour) [reply lib/build_runner.zig:759]: Maybe I'm missing the right vocabulary, but it is building and checking the graph (adding edges between nodes) and also creating the output stack, not _only_ checking for loops.

File: lib/build_runner.zig (hunk 2 above)
Comment (Sahnvour) [reply lib/build_runner.zig:380]: This is what I was going for at first, but I figured it would just be simpler for the user to accept any kind of data as argument, since you can't have bad formatting and your source of (random) seed can be pretty much anything (time[date] in any format, user or machine name, ...).
I'd expect the typical seed to be a number, whatever its format, but it does not have to be a hard requirement and the additional complexity is arguably very limited (a single hash over the bytes).

What do you think ?

**Comment by Sahnvour** [comment]
> Thanks for working on this!
> 
> Don't forget that part of the acceptance criteria is
> 
> > make the `zig build` subcommand always supply a random seed argument.

That's what
https://github.com/ziglang/zig/blob/615372aaa066e33bcb884b571f6323d7f8938161/lib/std/Build.zig#L271-L273
is about, am I missing something ?

File: lib/build_runner.zig (hunk 4 above)
Comment (Sahnvour) [reply lib/build_runner.zig:200]: Right I didn't realize this was the same scope but it's indeed better.

**Comment by andrewrk** [comment]
> am I missing something ?

Yeah- specifically the Zig CLI should choose the seed, so that when the build fails, the command to reproduce the issue includes the `--seed` argument.

File: lib/build_runner.zig (hunk 1 above)
Comment (rohlem) [reply lib/build_runner.zig:759]: Could maybe named `construct` to differentiate from "building" the project (by calling steps' `make` functions)? (If that's the ambiguity to avoid.)

File: lib/build_runner.zig (hunk 2 above)
Comment (andrewrk) [reply lib/build_runner.zig:380]: I think that hashing the seed is unnecessary. [I Want Simple, Not Just Easy](https://kristoff.it/blog/simple-not-just-easy/).

File: lib/build_runner.zig (hunk 3 above)
Comment (Sahnvour) [reply lib/build_runner.zig:1049]: This help message is ambiguous to me, is the default value supposed to be the seed used when the argument is not present, or when only `--seed` is passed with no value (and what would be the use case then ?) ?
Since the build runner is rarely called directly, but instead driven by `zig build`, should the zig cli mirror this if applicable ?

Comment (andrewrk) [reply lib/build_runner.zig:1049]: This usage text is the text printed by `zig build -h`, so it should be appropriate for those users. I made a mistake in the above suggestion. A better suggestion would be:

```
  --seed [integer]              Shuffles dependency traversal order (default: random)
```

The Zig CLI will choose a random seed if the user does not provide one explicitly. The build runner should use 0 if none is provided. However, in practice when using `zig build` a seed will always be provided.

Comment (Sahnvour) [reply lib/build_runner.zig:1049]: Okay, makes more sense thanks.

File: src/main.zig [hunk 6]
```diff
@@ -4389,6 +4368,8 @@ pub fn cmdBuild(gpa: Allocator, arena: Allocator, args: []const []const u8) !voi
         var child_argv = std.ArrayList([]const u8).init(arena);
         var reference_trace: ?u32 = null;
         var debug_compile_errors = false;
+        const micros: u32 = @truncate(@as(u64, @bitCast(std.time.microTimestamp())));
+        var seed: []const u8 = try std.fmt.allocPrint(arena, "{}", .{micros});
```
Comment (andrewrk) [inline src/main.zig:4372]: This is not how to seed a random number generator. You're involving a bunch of unrelated concepts such as the current time, memory allocation, and formatted string printing. Think about how unnecessarily complicated this is. And then at the end of this process, it didn't even succeed at the main goal, which was to obtain 32 random bits.

Anyway just call `std.crypto.random.int(u32)`.

**Review by andrewrk — CHANGES_REQUESTED** [review]
